javascript/ql/src/RegExp/IdentityReplacement.ql

```diff
@@ -0,0 +1,44 @@
+/**
+ * @name Replacement of a substring with itself
+ * @description Replacing a substring with itself has no effect and may indicate a mistake.
+ * @kind problem
+ * @problem.severity warning
+ * @id js/identity-replacement
+ * @tags correctness
+ * @precision very-high
+ */
+
+import javascript
+
+/**
+ * Holds if `e`, when used as the first argument of `String.prototype.replace`, matches
+ * `s` and nothing else.
+ */
+predicate matchesString(Expr e, string s) {
+  exists (RegExpConstant c |
+    matchesConstant(e.(RegExpLiteral).getRoot(), c) and
+    s = c.getValue()
+  )
+  or
+  s = e.getStringValue()
+}
+
+/**
+ * Holds if `t` matches `c` and nothing else.
+ */
+predicate matchesConstant(RegExpTerm t, RegExpConstant c) {
+  c = t
+  or
+  matchesConstant(t.(RegExpGroup).getAChild(), c)
+  or
+  exists (RegExpCharacterClass recc | recc = t and not recc.isInverted() |
+    recc.getNumChild() = 1 and
+    matchesConstant(recc.getChild(0), c)
+  )
```

Comment:
We should handle anchors for cases like `/^x$/`

Reply:
Good idea; and also other zero-width assertions like `\b`.